lib/travis/api/build/app.rb

```diff
@@ -37,6 +27,35 @@ class App < Sinatra::Base
           if ENV.key?('LIBRATO_EMAIL') && ENV.key?('LIBRATO_TOKEN') && ENV.key?('LIBRATO_SOURCE')
             use Metriks
           end
+
+          use Rack::Deflater
+        end
+
+        helpers do
+          def auth_disabled?
+            (
+              ENV['API_TOKEN'].nil? || ENV['API_TOKEN'].strip.empty?
+            ) && (
+              settings.development? || settings.testing?
+            )
+          end
+        end
+
+        before '/script' do
+          return if auth_disabled?
+
+          unless env.key?('HTTP_AUTHORIZATION')
+            halt 401, 'missing Authorization header'
+          end
+
+          type, token = env['HTTP_AUTHORIZATION'].to_s.split(' ', 2)
+
+          ENV['API_TOKEN'].split(',').each do |valid_token|
+            return if Rack::Utils.secure_compare(type, 'token') &&
+                      Rack::Utils.secure_compare(token, valid_token)
```

Comment:
It is "constant-time", but it will return fast if the length of the two strings are unequal, so you could figure out how long the actual token is by sending requests that have longer and longer tokens and check when there's a peak in response time due to it suddenly checking the content and not returning early. I don't leaking this information is necessarily bad, though, I think it's mostly noted in the docs for the method in case you don't want to leak that information.